feat: Implement 429 handling for request transport

kamilogorek opened the pull request:
1. Is this a good way to keep state?
2. How to make sure that `Agent` is always initialized in the module? L183 looks dubious.
3. How can I change server response based on the request number? Peeking at req body is meh.
4. Is there an easy way to mock time? Otherwise, 2 commented tests are really hard to implement without passing some time function around.

![image](https://user-images.githubusercontent.com/1523305/76741672-374cce80-6770-11ea-993e-7644b673f300.png)

kamilogorek commented on lib/sentry/client.ex:
```diff
@@ -186,18 +213,31 @@ defmodule Sentry.Client do
-         {:ok, json} <- json_library.decode(body) do
-      {:ok, Map.get(json, "id")}
-    else
+    case :hackney.request(:post, url, headers, body, hackney_opts) do
```
That's definitely not written with an optimal syntax

mitchellhenke commented:
@kamilogorek thanks for the pull request and sorry for the delay, I will be taking a look this week!

kamilogorek commented:
No rush @mitchellhenke , thanks!

mitchellhenke commented on lib/sentry/client.ex:
```diff
@@ -1,3 +1,23 @@
+  use Agent
+
+  def start_link do
+    Agent.start_link(fn -> Timex.now() end, name: __MODULE__)
```
Can replace with: `NaiveDateTime.utc_now()`

mitchellhenke commented on lib/sentry/client.ex:
```diff
@@ -387,6 +439,21 @@ defmodule Sentry.Client do
+      {:error, _} ->
+        try do
+          {retry_after, _} = Integer.parse(header, 10)
+          Timex.shift(Timex.now(), seconds: retry_after)
```
Can replace with:

```elixir
NaiveDateTime.add(NaiveDateTime.utc_now(), retry_after)
```

mitchellhenke commented on lib/sentry/client.ex:
```diff
@@ -1,3 +1,23 @@
+  end
+
+  def should_drop_event do
+    Agent.get(__MODULE__, fn state -> Timex.before?(Timex.now(), state) end)
```
Can replace with:
```NaiveDateTime.compare(NaiveDateTime.utc_now(), state) != :lt```

kamilogorek commented:
Thanks for a review! Updated code for agent and removed timex dependency (except for parsing). Unfortunately, I don't have too much spare time to work on this PR in near future so feel free to take it over if you're able to. Thanks!

mitchellhenke commented:
@kamilogorek thanks, I don't mind taking it from here!

kamilogorek commented:
Cleaning up my stale PRs. Feel free to just use branch directly. Thanks :)

mitchellhenke commented:
@kamilogorek thank you, I plan to get to this soon!
